src/platform/site-wide/side-nav/helpers.js

```diff
@@ -0,0 +1,94 @@
+// Dependencies
+import uuidv4 from 'uuid/v4';
+import { each, get, isEmpty } from 'lodash';
+
+/* Recursive function that derives `navItems` and `navItemsLookup` for `normalizeSideNavData`.
+  @param {Object}, options:
+  {
+    items: array,
+    parentID: string,
+    navItemsLookup: object,
+  }
+  @returns {Object}, navItemsLookup is keyed off `id`.
+*/
+const deriveNavItemsLookup = options => {
+  // Derive the properties from options.
+  const items = get(options, 'items', []);
+  const parentID = get(options, 'parentID');
+  let navItemsLookup = get(options, 'navItemsLookup', {});
+
+  each(items, (item, index) => {
+    // Derive item properties.
+    const description = get(item, 'description', '');
+    const expanded = get(item, 'expanded', false);
+    const href = get(item, 'url.path', '');
+    const id = uuidv4();
+    const label = get(item, 'label', '');
+    const nestedItems = get(item, 'links');
+
+    // Construct the formatted item.
+    const formattedItem = {
+      description,
+      expanded,
+      hasChildren: !isEmpty(nestedItems),
+      href,
+      id,
+      label,
+      order: index,
+      parentID,
+    };
+
+    // Always add the item to our lookup table keyed off by ID.
+    navItemsLookup[id] = formattedItem;
+
+    // Update our navItemsLookup with the nested items if the option is passed.
+    if (!isEmpty(nestedItems)) {
+      const nestedNavItemsLookup = deriveNavItemsLookup({
+        items: nestedItems,
+        navItemsLookup,
+        parentID: id,
+      });
+
+      // Update our navItemsLookup with the nested items.
+      navItemsLookup = {
+        ...navItemsLookup,
+        ...nestedNavItemsLookup,
+      };
+    }
+  });
+
+  return navItemsLookup;
```

Comment:
Hmm, can you think of a better name for it?

Reply:
Hm no the name is probably fine, but I'm wondering - what do think of using a `Map` for doing lookups instead of an object? No problem if that would be too much of a rewrite.

Reply:
Sure! That's a great idea 👍

Reply:
Putting this comment here for others who might come across this PR in the future:

MAP:
Benefits:
More expressive in that you have to call specific methods like get or set to interact with it.
Fits the navItemsLookup variable name better since Lookup Tables are often referred to as Lookup Maps.
Cons:
Very slight increase in memory use since it is effectively maintaining an array of kv pairs.
Other JS devs (e.g. who have worked in >ES5) may not be used to using it.
Do Lodash methods work with Maps?

OBJECT:
Benefits:
All JS devs are familiar with it (I hope lol).
Less memory use since it does not maintain insertion order.
Lodash methods definitely work with objects.
Cons:
Not as expressive as a Map.
Does not maintain insertion order (we'd need to create a separate variable to do that for it).


As a result, we ended up choosing an object for this over a Map.